File: python/letter_combinations_of_a_phone_number.py
numbers_to_letters = {
        "2": ["a", "b", "c"],
        "3": ["d", "e", "f"],
        "4": ["g", "h", "i"],
        "5": ["j", "k", "l"],
        "6": ["m", "n", "o"],
        "7": ["p", "q", "r", "s"],
        "8": ["t", "u", "v"],
        "9": ["w", "x", "y", "z"],
    }

class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        #keep track of the solutions as well as the stack outside of the recursion flow
        self.solutions = []
        self.digits = digits
        #store partial solution as a list rather than a string to allow constant time appending and popping
        self.stack = []
        self.backtrack()
        return self.solutions
        
    def backtrack(self) -> None:
        #if the stack is full, add it to the list of solutions
        if len(self.stack) == len(self.digits):
            if len(self.stack) > 0:
                self.solutions.append("".join(self.stack))
            return
        #otherwise backtrack for each possible next character
        for ch in numbers_to_letters[self.digits[len(self.stack)]]:
            self.stack.append(ch)
            self.backtrack()
            self.stack.pop()

File: java/letterCombinationsOfAPhoneNumber.java
class Solution {

    Map<Character, Character[]> numbersToLetters = new HashMap();
    //store stck as a StringBuilder to allow constant time insertion and removal
    StringBuilder stack = new StringBuilder();
    String digits = "";
    List<String> solutions = new ArrayList();
    
    public List<String> letterCombinations(String digits) {
        this.digits = digits;
        initializeMap();
        backtrack();
        return solutions;
    }
    
    private void backtrack() {
        //if the stack is currently full, add it to the list of solutions
        if (stack.length() == digits.length()) {
            if (stack.length() > 0) {
                solutions.add(stack.toString());
            }
            return;
        }
        //otherwise, call the backtrack function for each possible next character
        for (char c : numbersToLetters.get(digits.charAt(stack.length()))) {
            stack.append(c);
            backtrack();
            stack.deleteCharAt(stack.length() - 1);
        }
    }
    
    private void initializeMap() {
        numbersToLetters.put('2', new Character[]{ 'a', 'b', 'c' });
        numbersToLetters.put('3', new Character[]{ 'd', 'e', 'f' });
        numbersToLetters.put('4', new Character[]{ 'g', 'h', 'i' });
        numbersToLetters.put('5', new Character[]{ 'j', 'k', 'l' });
        numbersToLetters.put('6', new Character[]{ 'm', 'n', 'o' });
        numbersToLetters.put('7', new Character[]{ 'p', 'q', 'r', 's' });
        numbersToLetters.put('8', new Character[]{ 't', 'u', 'v' });
        numbersToLetters.put('9', new Character[]{ 'w', 'x', 'y', 'z' });
    }
}